azure-iot-device/azure/iot/device/iothub/pipeline/mqtt_pipeline.py

```diff
@@ -363,10 +437,22 @@ def enable_feature(self, feature_name, callback):
         :param callback: callback which is called when the feature is enabled
 
         :raises: ValueError if feature_name is invalid
+        :raises: :class:`azure.iot.device.iothub.pipeline.exceptions.PipelineNotRunning` if the
+            pipeline has previously been shut down
+
+        The following exceptions are not "raised", but rather returned via the "error" parameter
+        when invoking "callback":
+
+        :raises: :class:`azure.iot.device.iothub.pipeline.exceptions.ConnectionFailedError`
+        :raises: :class:`azure.iot.device.iothub.pipeline.exceptions.ConnectionDroppedError`
+        :raises: :class:`azure.iot.device.iothub.pipeline.exceptions.UnauthorizedError`
+        :raises: :class:`azure.iot.device.iothub.pipeline.exceptions.ProtocolClientError`
         """
+        self._ensure_running()
         logger.debug("enable_feature {} called".format(feature_name))
         if feature_name not in self.feature_enabled:
             raise ValueError("Invalid feature_name")
+        # TODO: What about if the feature is already enabled?
```

Comment:
># TODO: What about if the feature is already enabled? [](start = 8, length = 53)

This is an architectural question.  Enabling means subscribing, which means network traffic.  We want to make sure the feature is enabled, but we don't want to subscribe every time we use the feature.  We also don't want to assume that we're enabled when we're not.  

The real question is "what layer of code is responsible for making sure we don't subscribe more than we need to?"  Currently, this happens in sync_client.py, so I would argue that we should re-enable at this layer, even if we're already enabled.

For what it's worth, I'm currently dealing with a github issue that looks like a Python but because it doesn't repro with other languages because other languages will re-subscribe under certain conditions.  Right now, it looks like a bug in edgeHub, but there's some justification for following the rule: "re-subscribe if the user asks us, even if we think that we know better."

Reply:
that being said, I think changing the enable/disable code is out of scope and could be it's own little Pandora's box.

---
In reply to: [552299580](https://github.com/Azure/azure-iot-sdk-python/pull/721#discussion_r552299580) [](ancestors = 552299580)

Reply:
I agree. This was a little box I tripped over while doing this PR and didn't dare open, so I left this little TODO instead so it can't be forgotten about.